adapters/consumable/consumable.go

```diff
@@ -0,0 +1,292 @@
+}
+
+type bidResponse struct {
+	Decisions map[string]decision `json:"decisions"` // map by bidId
+}
+
+/**
+ * See https://dev.adzerk.com/v1.0/reference/response
+ */
+type decision struct {
+	Pricing       *pricing   `json:"pricing"`
+	AdID          int64      `json:"adId"`
+	BidderName    string     `json:"bidderName,omitempty"`
+	CreativeID    string     `json:"creativeId,omitempty"`
+	Contents      []contents `json:"contents"`
+	ImpressionUrl *string    `json:"impressionUrl,omitempty"`
+}
+
+type contents struct {
+	Body string `json:"body"`
+}
+
+type pricing struct {
+	ClearPrice *float64 `json:"clearPrice"`
+}
+
+func (a *ConsumableAdapter) MakeRequests(request *openrtb.BidRequest) ([]*adapters.RequestData, []error) {
+	headers := http.Header{
+		"Content-Type": {"application/json"},
+		"Accept":       {"application/json"},
+	}
+
+	if request.Device != nil {
+		if request.Device.UA != "" {
+			headers.Set("User-Agent", request.Device.UA)
+		}
+
+		if request.Device.IP != "" {
+			headers.Set("Forwarded", "for="+request.Device.IP)
+			headers.Set("X-Forwarded-For", request.Device.IP)
+		}
+	}
+
+	// Set azk cookie to one we got via sync
+	if request.User != nil {
+		userID := strings.TrimSpace(request.User.BuyerUID)
+		if len(userID) > 0 {
+			headers.Add("Cookie", fmt.Sprintf("%s=%s", "azk", userID))
+		}
+	}
+
+	if request.Site != nil && request.Site.Page != "" {
+		headers.Set("Referer", request.Site.Page)
+
+		pageUrl, err := url.Parse(request.Site.Page)
+		if err == nil {
+			origin := url.URL{
+				Scheme: pageUrl.Scheme,
+				Opaque: pageUrl.Opaque,
+				Host:   pageUrl.Host,
+			}
+
+			headers.Set("Origin", origin.String())
+		}
+	}
+
+	body := bidRequest{
+		Placements:         make([]placement, len(request.Imp)),
+		Time:               a.clock.Now().Unix(),
+		IncludePricingData: true,
+		EnableBotFiltering: true,
+		Parallel:           true,
+	}
+
+	if request.Site != nil {
+		body.Referrer = request.Site.Ref // Effectively the previous page to the page where the ad will be shown
+		body.Url = request.Site.Page     // where the impression will be made
+	}
+
+	for i, impression := range request.Imp {
+
+		_, consumableExt, err := extractExtensions(impression)
+
+		if err != nil {
+			return nil, err
+		}
+
+		// These get set on the first one in observed working requests
+		if i == 0 {
+			body.NetworkId = consumableExt.NetworkId
+			body.SiteId = consumableExt.SiteId
+			body.UnitId = consumableExt.UnitId
+			body.UnitName = consumableExt.UnitName
+		}
+
+		body.Placements[i] = placement{
+			DivName:   impression.ID,
+			NetworkId: consumableExt.NetworkId,
+			SiteId:    consumableExt.SiteId,
+			UnitId:    consumableExt.UnitId,
+			UnitName:  consumableExt.UnitName,
+			AdTypes:   getSizeCodes(impression.Banner.Format), // was adTypes: bid.adTypes || getSize(bid.sizes) in prebid.js
+		}
+	}
+
+	bodyBytes, err := json.Marshal(body)
+	if err != nil {
+		return nil, []error{err}
+	}
+
+	println("== header ==")
```

Comment:
Were these supposed to be `fmt.Printf()`s?

Reply:
Otherwise please only log to console messages that the PBS host needs to read in production. Else the logs become unreadable due to spam and useless for server maintenance.

Reply:
Are you referring to the println()?

Sorry. That should have been removed.  Will sort them.